File: libs/server/persistent/src/persistent/prisma/workspaces.ts
import { randomUUID } from 'node:crypto';
import {
  DomainError,
  Workspace,
  WorkspaceDescription,
  Workspaces,
} from '@evidence/server-domain';
import { EntityList } from '../database';
import {
  initializeWorkspaceModelRoot,
  publicWorkspaceMetadata,
} from '../workspace-paths';
import { assembleWorkspace } from './mappers';
import type { PrismaStore } from './types';
import { defaultIfBlank, inputJson, now } from './utils';

export class PrismaWorkspaces
  extends EntityList<Workspace>
  implements Workspaces
{
  constructor(
    private readonly store: PrismaStore,
    private readonly modelStorageRoot = process.env
      .EVIDENCE_WORKSPACE_STORAGE_ROOT,
  ) {
    super();
  }

  protected override async findEntities(
    from: number,
    to: number,
  ): Promise<Workspace[]> {
    const rows = await this.store.workspace.findMany({
      where: { deletedAt: null },
      orderBy: { updatedAt: 'desc' },
      skip: from,
      take: Math.max(to - from, 0),
    });
    return rows.map((row) => assembleWorkspace(this.store, row));
  }

  protected override async findEntity(id: string): Promise<Workspace | null> {
    const row = await this.store.workspace.findFirst({
      where: { id, deletedAt: null },
    });
    return row ? assembleWorkspace(this.store, row) : null;
  }

  override async size(): Promise<number> {
    return this.store.workspace.count({ where: { deletedAt: null } });
  }

  async create(
    ownerUserId: string,
    desc: WorkspaceDescription,
  ): Promise<Workspace> {
    const id = randomUUID();
    const timestamp = now();
    const metadata = publicWorkspaceMetadata(desc.metadata);
    const title = normalizeTitle(desc.title);
    const modelRoot = await initializeWorkspaceModelRoot(
      id,
      this.modelStorageRoot,
    );
    const createWorkspace = async (db: PrismaStore): Promise<Workspace> => {
      const row = await db.workspace.create({
        data: {
          id,
          title,
          description: desc.description,
          status: defaultIfBlank(desc.status, 'active'),
          metadata: inputJson(metadata),
          modelRoot,
          createdAt: timestamp,
          updatedAt: timestamp,
        },
      });

      await db.workspaceMember.create({
        data: {
          id: randomUUID(),
          workspaceId: id,
          userId: ownerUserId,
          role: 'owner',
          createdAt: timestamp,
          updatedAt: timestamp,
        },
      });

      return assembleWorkspace(this.store, row);
    };

    if ('$transaction' in this.store) {
      return this.store.$transaction((tx) => createWorkspace(tx));
    }
    return createWorkspace(this.store);
  }

  async update(id: string, desc: WorkspaceDescription): Promise<Workspace> {
    const current = await this.findByIdentity(id);
    if (!current) {
      throw DomainError.notFound(`workspace ${id} not found`);
    }

    const metadataInput =
      Object.keys(desc.metadata).length === 0
        ? current.description().metadata
        : desc.metadata;
    const metadata = publicWorkspaceMetadata(metadataInput);
    const row = await this.store.workspace.update({
      where: { id },
      data: {
        title: normalizeTitle(desc.title),
        description: desc.description,
        status: defaultIfBlank(desc.status, 'active'),
        metadata: inputJson(metadata),
        updatedAt: now(),
      },
    });
    return assembleWorkspace(this.store, row);
  }

  async delete(id: string): Promise<void> {
    const current = await this.findByIdentity(id);
    if (!current) {
      throw DomainError.notFound(`workspace ${id} not found`);
    }
    const timestamp = now();
    await this.store.workspace.update({
      where: { id },
      data: { deletedAt: timestamp, updatedAt: timestamp },
    });
  }
}

function normalizeTitle(title: string): string {
  const normalized = title.trim();
  if (!normalized) {
    throw DomainError.validation('workspace title must not be empty');
  }
  return normalized;
}

File: libs/server-java/persistent/src/main/java/reengineering/ddd/evidence/persistent/associations/Workspaces.java
package reengineering.ddd.evidence.persistent.associations;

import io.github.jayclock.smartdomain.core.Ref;
import io.github.jayclock.smartdomain.mybatis.database.EntityList;
import jakarta.inject.Inject;
import java.time.Clock;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.springframework.stereotype.Component;
import reengineering.ddd.evidence.domain.DomainException;
import reengineering.ddd.evidence.domain.description.MembershipDescription;
import reengineering.ddd.evidence.domain.description.WorkspaceDescription;
import reengineering.ddd.evidence.domain.model.Workspace;
import reengineering.ddd.evidence.persistent.WorkspaceModelRoot;
import reengineering.ddd.evidence.persistent.mappers.WorkspaceMembershipsMapper;
import reengineering.ddd.evidence.persistent.mappers.WorkspacesMapper;

@Component
public class Workspaces extends EntityList<String, Workspace>
    implements reengineering.ddd.evidence.domain.model.Workspaces {
  private final WorkspacesMapper mapper;
  private final WorkspaceMembershipsMapper membershipsMapper;
  private final WorkspaceModelRoot modelRoots;
  private final Clock clock;

  @Inject
  public Workspaces(
      WorkspacesMapper mapper,
      WorkspaceMembershipsMapper membershipsMapper,
      WorkspaceModelRoot modelRoots,
      Clock clock) {
    this.mapper = mapper;
    this.membershipsMapper = membershipsMapper;
    this.modelRoots = modelRoots;
    this.clock = clock;
  }

  @Override
  protected List<Workspace> findEntities(int from, int to) {
    return mapper.findAll(from, Math.max(to - from, 0));
  }

  @Override
  protected Workspace findEntity(String id) {
    return mapper.findByIdentity(id);
  }

  @Override
  public int size() {
    return mapper.countAll();
  }

  @Override
  public Workspace create(String ownerUserId, WorkspaceDescription description) {
    String id = UUID.randomUUID().toString();
    Instant timestamp = timestamp();
    WorkspaceDescription normalized = normalize(description, null, timestamp);
    mapper.insert(id, normalized, modelRoots.initializeWorkspace(id), timestamp);
    membershipsMapper.insert(
        UUID.randomUUID().toString(),
        id,
        new MembershipDescription(
            new Ref<>(id), new Ref<>(ownerUserId), "owner", timestamp, timestamp),
        timestamp);
    return require(id);
  }

  @Override
  public Workspace update(String workspaceId, WorkspaceDescription description) {
    Workspace current = require(workspaceId);
    WorkspaceDescription normalized = normalize(description, current.getDescription(), timestamp());
    if (mapper.update(workspaceId, normalized, normalized.updatedAt()) != 1) {
      throw DomainException.notFound("workspace " + workspaceId + " not found");
    }
    return require(workspaceId);
  }

  @Override
  public void delete(String workspaceId) {
    require(workspaceId);
    if (mapper.softDelete(workspaceId, timestamp()) != 1) {
      throw DomainException.notFound("workspace " + workspaceId + " not found");
    }
  }

  private Workspace require(String workspaceId) {
    return findByIdentity(workspaceId)
        .orElseThrow(() -> DomainException.notFound("workspace " + workspaceId + " not found"));
  }

  private WorkspaceDescription normalize(
      WorkspaceDescription input, WorkspaceDescription current, Instant timestamp) {
    String title = input.title() == null ? "" : input.title().trim();
    if (title.isEmpty()) {
      throw DomainException.validation("workspace title must not be empty");
    }
    String normalizedStatus = input.status() == null ? "" : input.status().trim();
    String status = normalizedStatus.isEmpty() ? "active" : normalizedStatus;
    Map<String, String> metadata =
        input.metadata().isEmpty() && current != null ? current.metadata() : input.metadata();
    return new WorkspaceDescription(
        title,
        input.description(),
        status,
        modelRoots.publicMetadata(metadata),
        current == null ? timestamp : current.createdAt(),
        timestamp);
  }

  private Instant timestamp() {
    return clock.instant().truncatedTo(ChronoUnit.MILLIS);
  }
}
